File: astenn-core/src/main/java/org/lestr/astenn/configuration/CompositePersistenceDriver.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package org.lestr.astenn.configuration;

import org.lestr.astenn.plugin.IPersistenceDriver;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

/**
 *
 * @author PIBONNIN
 */
public class CompositePersistenceDriver implements IPersistenceDriver {


    private IPersistenceDriver readWritePersistenceDriver;


    private Collection<IPersistenceDriver> readOnlyPersistenceDrivers;


    public CompositePersistenceDriver() {

        readWritePersistenceDriver = new RAMPersistenceDriver();
        readOnlyPersistenceDrivers = new ArrayList<IPersistenceDriver>();

    }// END Constructor


    public CompositePersistenceDriver(IPersistenceDriver readWritePersistenceDriver,
                                      IPersistenceDriver... readOnlyPersistenceDrivers) {

        this.readWritePersistenceDriver = readWritePersistenceDriver;
        this.readOnlyPersistenceDrivers = new ArrayList<IPersistenceDriver>();

        this.readOnlyPersistenceDrivers.addAll(Arrays.asList(readOnlyPersistenceDrivers));

    }// END Constructor


    /**
     * @return the ReadWritePersistenceDriver
     */
    public IPersistenceDriver getReadWritePersistenceDriver() {

        return readWritePersistenceDriver;

    }// END Method getReadWritePersistenceDriver


    /**
     * @param ReadWritePersistenceDriver the ReadWritePersistenceDriver to set
     */
    public void setReadWritePersistenceDriver(IPersistenceDriver ReadWritePersistenceDriver) {

        this.readWritePersistenceDriver = ReadWritePersistenceDriver;

    }// END Method setReadWritePersistenceDriver


    /**
     * @return the ReadOnlyPersistenceDrivers
     */
    public Collection<IPersistenceDriver> getReadOnlyPersistenceDrivers() {

        return readOnlyPersistenceDrivers;

    }// END Method getReadOnlyPersistenceDrivers


    @Override
    public void addPluginInterface(String pluginInterfaceName) {

        readWritePersistenceDriver.addPluginInterface(pluginInterfaceName);

    }// END Method addPluginInterface


    @Override
    public void removePluginInterface(String pluginInterfaceName) {

        readWritePersistenceDriver.removePluginInterface(pluginInterfaceName);

    }// END Method removePluginInterface


    @Override
    public boolean existPluginInterface(String pluginInterfaceName) {

        if (readWritePersistenceDriver.existPluginInterface(pluginInterfaceName))
            return true;

        for (IPersistenceDriver driver : readOnlyPersistenceDrivers)
            if (driver.existPluginInterface(pluginInterfaceName))
                return true;

        return false;

    }// END Method existPluginInterface


    @Override
    public void addPluginImplementation(String pluginInterfaceName,
                                        String pluginImplementationAddress) {

        readWritePersistenceDriver.addPluginImplementation(pluginInterfaceName, pluginImplementationAddress);

    }// END Method addPluginImplementation


    @Override
    public void removePluginImplementation(String pluginInterfaceName,
                                           String pluginImplementationAddress) {

        if (readWritePersistenceDriver.existPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
            readWritePersistenceDriver.removePluginImplementation(pluginInterfaceName, pluginImplementationAddress);

    }// END Method removePluginImplementation


    @Override
    public boolean existPluginImplementation(String pluginInterfaceName,
                                             String pluginImplementationAddress) {

        if (readWritePersistenceDriver.existPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
            return true;

        for (IPersistenceDriver driver : readOnlyPersistenceDrivers)
            if (driver.existPluginInterface(pluginInterfaceName)
                && driver.existPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
                return true;

        return false;

    }// END Method existPluginImplementation


    @Override
    public Iterable<String> getPluginInterfacesNames() {

        ArrayList<String> rslt = new ArrayList<String>();

        for (String interfaceName : readWritePersistenceDriver.getPluginInterfacesNames())
            rslt.add(interfaceName);

        for (IPersistenceDriver driver : readOnlyPersistenceDrivers)
            for (String interfaceName : driver.getPluginInterfacesNames())
                if (!rslt.contains(interfaceName))
                    rslt.add(interfaceName);

        return rslt;

    }// END Method getPluginInterfacesNames


    @Override
    public Iterable<String> getPluginImplementationsAddresses(String pluginInterfaceName) {

        ArrayList<String> rslt = new ArrayList<String>();

        if (readWritePersistenceDriver.existPluginInterface(pluginInterfaceName))
            for (String implementation : readWritePersistenceDriver.getPluginImplementationsAddresses(pluginInterfaceName))
                rslt.add(implementation);

        for (IPersistenceDriver driver : readOnlyPersistenceDrivers)
            if (driver.existPluginInterface(pluginInterfaceName))
                for (String implementation : driver.getPluginImplementationsAddresses(pluginInterfaceName))
                    if (!rslt.contains(implementation))
                        rslt.add(implementation);

        return rslt;

    }// END Method getPluginImplementationsStrings


}// END Class CompositePersistenceDriver

File: dotNet/astenn/astenn-core/src/Org/Lestr/Astenn/Configuration/CompositePersistenceDriver.cs
/*
 * Crée par SharpDevelop.
 * Utilisateur: Choupis
 * Date: 30/04/2010
 * Heure: 22:33
 * 
 * Pour changer ce modèle utiliser Outils | Options | Codage | Editer les en-têtes standards.
 */
using System;
using System.Collections.Generic;
using Org.Lestr.Astenn.Plugin;

namespace Org.Lestr.Astenn.Configuration
{
	
	
	public class CompositePersistenceDriver : IPersistenceDriver {
	
	
	
	    private IPersistenceDriver readWritePersistenceDriver;
	    private List<IPersistenceDriver> readOnlyPersistenceDrivers;
	
	
	    public CompositePersistenceDriver()
	    {
	
	        readWritePersistenceDriver = new RAMPersistenceDriver();
	        readOnlyPersistenceDrivers = new List<IPersistenceDriver>();
	
	    }// END Constructor
	
	
	    public CompositePersistenceDriver(IPersistenceDriver readWritePersistenceDriver, IEnumerable<IPersistenceDriver> readOnlyPersistenceDrivers)
	    {

            this.readWritePersistenceDriver = readWritePersistenceDriver;
            this.readOnlyPersistenceDrivers = new List<IPersistenceDriver>(readOnlyPersistenceDrivers);
	
	        foreach(IPersistenceDriver driver in readOnlyPersistenceDrivers)
	            this.readOnlyPersistenceDrivers.Add(driver);
	
	    }// END Constructor
	    
	    
		public List<IPersistenceDriver> ReadOnlyPersistenceDrivers {
	    	
			get { return readOnlyPersistenceDrivers; }
			
		}// END Property ReadOnlyPersistenceDrivers
	    
	    
		public IPersistenceDriver ReadWritePersistenceDriver {
	    	
			get { return readWritePersistenceDriver; }
			set { readWritePersistenceDriver = value; }
			
		}// END Property ReadWritePersistenceDriver
	
	
	    public void AddPluginInterface(string pluginInterfaceName) {
	
	        readWritePersistenceDriver.AddPluginInterface(pluginInterfaceName);
	
	    }// END Property AddPluginInterface
	
	
	    public void RemovePluginInterface(string pluginInterfaceName) {
	
			if(readWritePersistenceDriver.ExistPluginInterface(pluginInterfaceName))
				readWritePersistenceDriver.RemovePluginInterface(pluginInterfaceName);
	
	    }// END Property RemovePluginInterface
	
	
	    public bool ExistPluginInterface(string pluginInterfaceName) {
	
	        if(readWritePersistenceDriver.ExistPluginInterface(pluginInterfaceName))
	            return true;
	
	        foreach(IPersistenceDriver driver in readOnlyPersistenceDrivers)
	            if(driver.ExistPluginInterface(pluginInterfaceName))
	                return true;
	
	        return false;
	
	    }// END Property ExistPluginInterface


        public void AddPluginImplementation(string pluginInterfaceName, string pluginImplementationAddress)
        {
	
	        readWritePersistenceDriver.AddPluginImplementation(pluginInterfaceName, pluginImplementationAddress);
	
	    }// END Property AddPluginImplementation


        public void RemovePluginImplementation(string pluginInterfaceName, string pluginImplementationAddress)
        {

			if (readWritePersistenceDriver.ExistPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
				readWritePersistenceDriver.RemovePluginImplementation(pluginInterfaceName, pluginImplementationAddress);
	
	    }// END Property RemovePluginImplementation


        public bool ExistPluginImplementation(string pluginInterfaceName, string pluginImplementationAddress)
        {

            if (readWritePersistenceDriver.ExistPluginInterface(pluginInterfaceName)
                && readWritePersistenceDriver.ExistPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
	            return true;
	
	        foreach(IPersistenceDriver driver in readOnlyPersistenceDrivers)
	            if(driver.ExistPluginInterface(pluginInterfaceName)
                    && driver.ExistPluginImplementation(pluginInterfaceName, pluginImplementationAddress))
	                return true;
	        
	        return false;
	
	    }// END Property ExistPluginImplementation
	
	
	    public IEnumerable<string> GetPluginInterfacesNames() {
	
	        List<String> rslt = new List<String>();
	
	        foreach(string interfaceName in readWritePersistenceDriver.GetPluginInterfacesNames())
	            rslt.Add(interfaceName);
	
	        foreach(IPersistenceDriver driver in readOnlyPersistenceDrivers)
	            foreach(string interfaceName in driver.GetPluginInterfacesNames())
	                if(!rslt.Contains(interfaceName))
	                    rslt.Add(interfaceName);
	
	        return rslt;
	
	    }// END Method GetPluginImplementationsNames


        public IEnumerable<string> GetPluginImplementationsAddresses(string pluginInterfaceName)
        {

            List<string> rslt = new List<string>();

            if (readWritePersistenceDriver.ExistPluginInterface(pluginInterfaceName))
                foreach (string implementation in readWritePersistenceDriver.GetPluginImplementationsAddresses(pluginInterfaceName))
	                rslt.Add(implementation);
	
	        foreach(IPersistenceDriver driver in readOnlyPersistenceDrivers)
                if (driver.ExistPluginInterface(pluginInterfaceName))
                    foreach (string implementation in driver.GetPluginImplementationsAddresses(pluginInterfaceName))
	                    if(!rslt.Contains(implementation))
	                        rslt.Add(implementation);
	
	        return rslt;
	
	    }// END Property GetPluginImplementationsNames
	
	
	}// END Class CompositePersistenceDriver
	
	
}// END Namespace Astenn.Configuration
